src/librbd/TaskFinisher.h

```diff
@@ -71,17 +71,22 @@ class TaskFinisher {
     return true;
   }
 
-  void cancel_all(Context *comp) {
-    {
-      std::lock_guard l{*m_lock};
-      for (typename TaskContexts::iterator it = m_task_contexts.begin();
-           it != m_task_contexts.end(); ++it) {
+  template <typename L>
+  void cancel_if(L &&test, std::vector<Task> *canceled_tasks = nullptr) {
```

Comment:
Nit: The lambda could just as easily build a `canceled_tasks` vector instead of passing the responsibility to `cancel_if`.